impeller/entity/geometry/stroke_path_geometry.cc

```diff
@@ -38,225 +82,312 @@ Join StrokePathGeometry::GetStrokeJoin() const {
+                  const StrokePathGeometry::JoinProc& p_join_proc,
+                  const StrokePathGeometry::CapProc& p_cap_proc,
+                  const Scalar p_scale,
+                  VertexWriter& p_vtx_builder)
+      : polyline(p_polyline),
+        stroke_width(p_stroke_width),
+        scaled_miter_limit(p_scaled_miter_limit),
+        join_proc(p_join_proc),
+        cap_proc(p_cap_proc),
+        scale(p_scale),
+        vtx_builder(p_vtx_builder) {}
+
+  void Generate() {
+    for (size_t contour_i = 0; contour_i < polyline.contours.size();
+         contour_i++) {
+      auto contour = polyline.contours[contour_i];
+      size_t contour_start_point_i, contour_end_point_i;
+      std::tie(contour_start_point_i, contour_end_point_i) =
+          polyline.GetContourPointBounds(contour_i);
+
+      switch (contour_end_point_i - contour_start_point_i) {
+        case 1: {
+          Point p = polyline.GetPoint(contour_start_point_i);
+          cap_proc(vtx_builder, p, {-stroke_width * 0.5f, 0}, scale, false);
+          cap_proc(vtx_builder, p, {stroke_width * 0.5f, 0}, scale, false);
+          continue;
+        }
+        case 0:
+          continue;  // This contour has no renderable content.
+        default:
+          break;
+      }
 
-  VS::PerVertexData vtx;
+      ComputeOffset(contour_start_point_i, contour_start_point_i,
+                    contour_end_point_i, contour);
+      const Point contour_first_offset = offset;
+
+      if (contour_i > 0) {
+        // This branch only executes when we've just finished drawing a contour
+        // and are switching to a new one.
+        // We're drawing a triangle strip, so we need to "pick up the pen" by
+        // appending two vertices at the end of the previous contour and two
+        // vertices at the start of the new contour (thus connecting the two
+        // contours with two zero volume triangles, which will be discarded by
+        // the rasterizer).
+        vtx.position = polyline.GetPoint(contour_start_point_i - 1);
+        // Append two vertices when "picking up" the pen so that the triangle
+        // drawn when moving to the beginning of the new contour will have zero
+        // volume.
+        vtx_builder.AppendVertex(vtx.position);
+        vtx_builder.AppendVertex(vtx.position);
+
+        vtx.position = polyline.GetPoint(contour_start_point_i);
+        // Append two vertices at the beginning of the new contour, which
+        // appends  two triangles of zero area.
+        vtx_builder.AppendVertex(vtx.position);
+        vtx_builder.AppendVertex(vtx.position);
+      }
 
-  // Offset state.
-  Point offset;
-  Point previous_offset;  // Used for computing joins.
+      // Generate start cap.
+      if (!polyline.contours[contour_i].is_closed) {
+        auto cap_offset =
+            Vector2(-contour.start_direction.y, contour.start_direction.x) *
+            stroke_width * 0.5;  // Counterclockwise normal
```

Comment:
nit: I'm pretty sure you are incurring the cost of converting back and forth from 64bit floats to 32bit floats.  This should probably be `0.5f`.

Reply:
Done